src/tools/kata-ctl/src/arch/x86_64/mod.rs

```diff
@@ -141,6 +168,100 @@ mod arch_specific {
 
         Ok(GuestProtection::NoProtection)
     }
+
+    fn running_on_vmm() -> Result<bool> {
+        match check::get_single_cpu_info(check::PROC_CPUINFO, CPUINFO_DELIMITER) {
+            Ok(cpu_info) => {
+                // check if the 'hypervisor' flag exist in the cpu features
+                let missing_hypervisor_flag = check::check_cpu_attribs(&cpu_info, VMM_FLAGS)?;
+
+                if missing_hypervisor_flag.is_empty() {
+                    return Ok(true);
+                }
+            }
+            Err(e) => {
+                return Err(anyhow!("{}: {}", e, check::PROC_CPUINFO));
+            }
+        }
+
+        Ok(false)
+    }
+
+    fn unrestricted_guest_param_check(current_value: &str) -> Result<()> {
+        let expected_param_value: char = 'Y';
+
+        match running_on_vmm() {
+            Ok(running_on_vmm) => {
+                if !running_on_vmm {
+                    // check the value of the parameter and check if we are running inside a VMM
+                    if current_value == expected_param_value.to_string() {
+                        return Ok(());
+                    } else {
+                        return Err(anyhow!(
+                            "{} does not meet the expected value {}",
```

Comment:
@c3d - I agree this does look a bit confusing. However, when this error propagates back, the user will end up seeing something like this (see `check_kernel_modules()`):

```
ERROR: "Kernel Module: kvm Parameter: kvmclock_periodic_sync = Y does not meet the expected value: N"
```

That contains all the detail, but taking another look I am wondering if a better approach would be to not "chain" together errors like this and instead change the definition of `BuiltinModuleParamFp`:

```diff
diff --git a/src/tools/kata-ctl/src/types.rs b/src/tools/kata-ctl/src/types.rs
index f71daece0..26f5954d8 100644
--- a/src/tools/kata-ctl/src/types.rs
+++ b/src/tools/kata-ctl/src/types.rs
@@ -35,8 +35,18 @@ pub struct CheckItem<'a> {
     pub perm: PermissionType,
 }

-// Builtin module parameter check handler type
-pub type BuiltinModuleParamFp = fn(args: &str) -> Result<()>;
+// Builtin module parameter check handler type.
+//
+// BuiltinModuleParamFp represents a predicate function to determine if a
+// kernel parameter _value_ is as expected. If not, the returned Error will
+// explain what is wrong.
+//
+// Parameters:
+//
+// - module: name of kernel module.
+// - param: name of parameter for the kernel module.
+// - value: value of the kernel parameter.
+pub type BuiltinModuleParamFp = fn(module: &str, param: &str, value: &str) -> Result<()>;

 // KernelParamType encodes the value and a handler
 // function for kernel module parameters
```

This way, the BuiltinModuleParamFp can return the full error message (containing the kernel module and kernel param name. That would also simplify  `check_kernel_modules()` as it could just call:

```diff
diff --git a/src/tools/kata-ctl/src/arch/x86_64/mod.rs b/src/tools/kata-ctl/src/arch/x86_64/mod.rs
index c2501cbef..e8a059c75 100644
--- a/src/tools/kata-ctl/src/arch/x86_64/mod.rs
+++ b/src/tools/kata-ctl/src/arch/x86_64/mod.rs
@@ -243,15 +243,7 @@ mod arch_specific {

                     match parameter_check {
                         Ok(_v) => println!("{} Ok", module.name),
-
-                        Err(parameter_status) => {
-                            return Err(anyhow!(
-                                "Kernel Module: {} Parameter: {:} = {:}",
-                                module.name,
-                                module.parameter.name.to_string(),
-                                parameter_status,
-                            ));
-                        }
+                        Err(e) => return Err(e),
                     }
                 }
```

I'd also tweak the format slightly. Maybe something like:

```
ERROR: Kernel module 'kvm': parameter 'kvmclock_periodic_sync' should have value 'Y', but found 'N'.
```

To be super helpful, we could explain how the user/admin might be able to resolve the problem. Something like:

```
ERROR: Kernel module 'kvm': parameter 'kvmclock_periodic_sync' should have value 'Y', but found 'N'.
 (try running "sudo modprobe -r kvm ; sudo modprobe kvm kvmclock_periodic_sync=Y")
```

However, those commands may not work due to module deps so maybe we should just say, something like:

```
ERROR: Kernel module 'kvm': parameter 'kvmclock_periodic_sync' should have value 'Y', but found 'N': remove the 'kvm' module using `rmmod` and then reload using `modprobe`, setting 'kvmclock_periodic_sync=Y' ?
```

> **Note:**
>
> Although GitHub formats them across multiple lines, I think all these errors should be on a single-line.

Reply:
@c3d I appreciate your suggestion to improve the error messages and propagate the error from just one source.
@jodh-intel thank you for this better approach to describe the error.

Reply:
@dborquez Do you mind doing another iteration with @jodh-intel's suggestions?